File: ts/packages/luzid-grpc-client/src/client/workspace.ts
import { createClient } from '@luzid/grpc-connection'
import type { Channel } from '@luzid/grpc-connection'

import {
  GetWorkspaceRequest,
  GetWorkspaceResponse,
  CloneWorkspaceRequest,
  CloneWorkspaceResponse,
  WatchWorkspaceRequest,
  WatchWorkspaceResponse,
  UnwatchWorkspaceRequest,
  UnwatchWorkspaceResponse,
  AddWorkspaceRequest,
  AddWorkspaceResponse,
  RemoveWorkspaceRequest,
  RemoveWorkspaceResponse,
  ListWorkspacesResponse,
  WorkspaceServiceClient,
  WorkspaceServiceDefinition,
} from '@luzid/grpc'

export {
  GetWorkspaceRequest,
  GetWorkspaceResponse,
  CloneWorkspaceRequest,
  CloneWorkspaceResponse,
  WatchWorkspaceRequest,
  WatchWorkspaceResponse,
  UnwatchWorkspaceRequest,
  UnwatchWorkspaceResponse,
  AddWorkspaceRequest,
  AddWorkspaceResponse,
  RemoveWorkspaceRequest,
  RemoveWorkspaceResponse,
  ListWorkspacesResponse,
  WorkspaceProgram,
  Workspace,
  ClonableWorkspaceProgram,
  ClonableWorkspace,
} from '@luzid/grpc'

// -----------------
// WorkspaceGetClient
// -----------------
class WorkspaceGetClient {
  private readonly client: WorkspaceServiceClient

  constructor(channel: Channel) {
    this.client = createClient(WorkspaceServiceDefinition, channel)
  }

  getWorkspace(request: GetWorkspaceRequest): Promise<GetWorkspaceResponse> {
    return this.client.getWorkspace(request)
  }
}

// -------------------
// WorkspaceCloneClient
// -------------------
class WorkspaceCloneClient {
  private readonly client: WorkspaceServiceClient

  constructor(channel: Channel) {
    this.client = createClient(WorkspaceServiceDefinition, channel)
  }

  cloneWorkspace(
    request: CloneWorkspaceRequest
  ): Promise<CloneWorkspaceResponse> {
    return this.client.cloneWorkspace(request)
  }
}

// ---------------------
// WorkspaceWatchClient
// ---------------------
class WorkspaceWatchClient {
  private readonly client: WorkspaceServiceClient

  constructor(channel: Channel) {
    this.client = createClient(WorkspaceServiceDefinition, channel)
  }

  watchWorkspace(
    request: WatchWorkspaceRequest
  ): Promise<WatchWorkspaceResponse> {
    return this.client.watchWorkspace(request)
  }

  unwatchWorkspace(
    request: UnwatchWorkspaceRequest
  ): Promise<UnwatchWorkspaceResponse> {
    return this.client.unwatchWorkspace(request)
  }
}

// -------------------
// WorkspacePersistClient
// -------------------
class WorkspacePersistClient {
  private readonly client: WorkspaceServiceClient

  constructor(channel: Channel) {
    this.client = createClient(WorkspaceServiceDefinition, channel)
  }

  addWorkspace(request: AddWorkspaceRequest): Promise<AddWorkspaceResponse> {
    return this.client.addWorkspace(request)
  }

  removeWorkspace(
    request: RemoveWorkspaceRequest
  ): Promise<RemoveWorkspaceResponse> {
    return this.client.removeWorkspace(request)
  }

  listWorkspaces(): Promise<ListWorkspacesResponse> {
    return this.client.listWorkspaces({})
  }
}

// -----------------
// Consolidated WorkspaceClient
// -----------------
export class WorkspaceClient {
  private readonly getWorkspaceClient: WorkspaceGetClient
  private readonly cloneWorkspaceClient: WorkspaceCloneClient
  private readonly watchWorkspaceClient: WorkspaceWatchClient
  private readonly persistWorkspaceClient: WorkspacePersistClient

  constructor(channel: Channel) {
    this.getWorkspaceClient = new WorkspaceGetClient(channel)
    this.cloneWorkspaceClient = new WorkspaceCloneClient(channel)
    this.watchWorkspaceClient = new WorkspaceWatchClient(channel)
    this.persistWorkspaceClient = new WorkspacePersistClient(channel)
  }

  getWorkspace(request: GetWorkspaceRequest): Promise<GetWorkspaceResponse> {
    return this.getWorkspaceClient.getWorkspace(request)
  }

  cloneWorkspace(
    request: CloneWorkspaceRequest
  ): Promise<CloneWorkspaceResponse> {
    return this.cloneWorkspaceClient.cloneWorkspace(request)
  }

  watchWorkspace(
    request: WatchWorkspaceRequest
  ): Promise<WatchWorkspaceResponse> {
    return this.watchWorkspaceClient.watchWorkspace(request)
  }

  unwatchWorkspace(
    request: UnwatchWorkspaceRequest
  ): Promise<UnwatchWorkspaceResponse> {
    return this.watchWorkspaceClient.unwatchWorkspace(request)
  }

  addWorkspace(request: AddWorkspaceRequest): Promise<AddWorkspaceResponse> {
    return this.persistWorkspaceClient.addWorkspace(request)
  }

  removeWorkspace(
    request: RemoveWorkspaceRequest
  ): Promise<RemoveWorkspaceResponse> {
    return this.persistWorkspaceClient.removeWorkspace(request)
  }

  listWorkspaces(): Promise<ListWorkspacesResponse> {
    return this.persistWorkspaceClient.listWorkspaces()
  }
}

File: dart/luzid-grpc-client/lib/src/client/workspace.dart
import 'package:grpc/grpc_connection_interface.dart';
import 'package:luzid_grpc/luzid_grpc.dart';

// -----------------
// WorkspaceGetClient
// -----------------
class WorkspaceGetClient {
  final WorkspaceServiceClient _client;

  WorkspaceGetClient(ClientChannelBase channel)
      : _client = WorkspaceServiceClient(channel);

  Future<GetWorkspaceResponse> getWorkspace(GetWorkspaceRequest request) {
    return _client.getWorkspace(request);
  }
}

// -------------------
// WorkspaceCloneClient
// -------------------
class WorkspaceCloneClient {
  final WorkspaceServiceClient _client;

  WorkspaceCloneClient(ClientChannelBase channel)
      : _client = WorkspaceServiceClient(channel);

  Future<CloneWorkspaceResponse> cloneWorkspace(
    CloneWorkspaceRequest request,
  ) {
    return _client.cloneWorkspace(request);
  }
}

// ---------------------
// WorkspaceWatchClient
// ---------------------
class WorkspaceWatchClient {
  final WorkspaceServiceClient _client;

  WorkspaceWatchClient(ClientChannelBase channel)
      : _client = WorkspaceServiceClient(channel);

  Future<WatchWorkspaceResponse> watchWorkspace(
    WatchWorkspaceRequest request,
  ) {
    return _client.watchWorkspace(request);
  }

  Future<UnwatchWorkspaceResponse> unwatchWorkspace(
    UnwatchWorkspaceRequest request,
  ) {
    return _client.unwatchWorkspace(request);
  }
}

// -------------------
// WorkspacePersistClient
// -------------------
class WorkspacePersistClient {
  final WorkspaceServiceClient _client;

  WorkspacePersistClient(ClientChannelBase channel)
      : _client = WorkspaceServiceClient(channel);

  Future<AddWorkspaceResponse> addWorkspace(AddWorkspaceRequest request) {
    return _client.addWorkspace(request);
  }

  Future<RemoveWorkspaceResponse> removeWorkspace(
    RemoveWorkspaceRequest request,
  ) {
    return _client.removeWorkspace(request);
  }

  Future<ListWorkspacesResponse> listWorkspaces() {
    return _client.listWorkspaces(Empty());
  }
}

// -----------------
// WorkspaceWatchChangedClient
// -----------------
class WorkspaceWatchChangedClient {
  final WorkspaceWatchChangedSubClient _client;

  WorkspaceWatchChangedClient(ClientChannelBase channel)
      : _client = WorkspaceWatchChangedSubClient(channel);

  Stream<WorkspaceWatchChanged> subWorkspaceWatchChanged() {
    return _client.subWorkspaceWatchChanged(Empty());
  }
}

// -----------------
// Consolidated WorkspaceClient
// -----------------
class WorkspaceClient {
  final WorkspaceGetClient _getWorkspaceClient;
  final WorkspaceCloneClient _cloneWorkspaceClient;
  final WorkspaceWatchClient _watchWorkspaceClient;
  final WorkspacePersistClient _persistWorkspaceClient;
  final WorkspaceWatchChangedClient _workspaceWatchChangedClient;

  WorkspaceClient(ClientChannelBase channel)
      : _getWorkspaceClient = WorkspaceGetClient(channel),
        _cloneWorkspaceClient = WorkspaceCloneClient(channel),
        _watchWorkspaceClient = WorkspaceWatchClient(channel),
        _persistWorkspaceClient = WorkspacePersistClient(channel),
        _workspaceWatchChangedClient = WorkspaceWatchChangedClient(channel);

  Future<GetWorkspaceResponse> getWorkspace(GetWorkspaceRequest request) {
    return _getWorkspaceClient.getWorkspace(request);
  }

  Future<CloneWorkspaceResponse> cloneWorkspace(
    CloneWorkspaceRequest request,
  ) {
    return _cloneWorkspaceClient.cloneWorkspace(request);
  }

  Future<WatchWorkspaceResponse> watchWorkspace(
    WatchWorkspaceRequest request,
  ) {
    return _watchWorkspaceClient.watchWorkspace(request);
  }

  Future<UnwatchWorkspaceResponse> unwatchWorkspace(
    UnwatchWorkspaceRequest request,
  ) {
    return _watchWorkspaceClient.unwatchWorkspace(request);
  }

  Future<AddWorkspaceResponse> addWorkspace(AddWorkspaceRequest request) {
    return _persistWorkspaceClient.addWorkspace(request);
  }

  Future<RemoveWorkspaceResponse> removeWorkspace(
    RemoveWorkspaceRequest request,
  ) {
    return _persistWorkspaceClient.removeWorkspace(request);
  }

  Future<ListWorkspacesResponse> listWorkspaces() {
    return _persistWorkspaceClient.listWorkspaces();
  }

  Stream<WorkspaceWatchChanged> subWorkspaceWatchChanged() {
    return _workspaceWatchChangedClient.subWorkspaceWatchChanged();
  }
}
